### PR TITLE
hack/usoc22: Update session 01 - Baby Steps

### PR DESCRIPTION
**Pull request by eduardvintila**
Signed-off-by: Eduard Vintilă <eduard.vintila47@gmail.com>

### CODE REVIEW

File: content/en/community/hackathons/sessions/baby-steps/content/Demos_01.-Building-and-Running-the-Helloworld-Application.md [hunk 1]
```diff
@@ -7,12 +7,12 @@ The `kraft up` command makes it easy to do that with one swoop.
 Let's do that for the `helloworld` application (listed with `kraft list`):
 ```
 $ kraft up -t helloworld hello
- 100.00% :::::::::::::::::::::::::::::::::::::::: |       21 /       21 |:  app/helloworld@0.5
-[INFO    ] Initialized new unikraft application: /home/razvan/hello
-make: Entering directory '/home/razvan/.unikraft/unikraft'
+ 100.00% :::::::::::::::::::::::::::::::::::::::: |       21 /       21 |:  app/helloworld@0.10.0
```
Comment (RaduNichita) [inline content/en/community/hackathons/sessions/baby-steps/content/Demos_01.-Building-and-Running-the-Helloworld-Application.md:10]: I think we should rename these files to by demo_XY_short_description

File: content/en/community/hackathons/sessions/baby-steps/content/Demos_01.-Building-and-Running-the-Helloworld-Application.md [hunk 2]
```diff
@@ -7,12 +7,12 @@ The `kraft up` command makes it easy to do that with one swoop.
...
+[INFO    ] Initialized new unikraft application: /home/X/hello
+make: Entering directory '/home/X/.unikraft/unikraft'
```
Comment (RaduNichita) [inline content/en/community/hackathons/sessions/baby-steps/content/Demos_01.-Building-and-Running-the-Helloworld-Application.md:12]:
```suggestion
make: Entering directory '/home/unikraft/.unikraft/unikraft'
```

File: content/en/community/hackathons/sessions/baby-steps/content/Demos_01.-Building-and-Running-the-Helloworld-Application.md [hunk 3]
```diff
@@ -7,12 +7,12 @@ The `kraft up` command makes it easy to do that with one swoop.
 Let's do that for the `helloworld` application (listed with `kraft list`):
 ```
 $ kraft up -t helloworld hello
- 100.00% :::::::::::::::::::::::::::::::::::::::: |       21 /       21 |:  app/helloworld@0.5
-[INFO    ] Initialized new unikraft application: /home/razvan/hello
-make: Entering directory '/home/razvan/.unikraft/unikraft'
+ 100.00% :::::::::::::::::::::::::::::::::::::::: |       21 /       21 |:  app/helloworld@0.10.0
+[INFO    ] Initialized new unikraft application: /home/X/hello
```
Comment (RaduNichita) [inline content/en/community/hackathons/sessions/baby-steps/content/Demos_01.-Building-and-Running-the-Helloworld-Application.md:11]: I think we should replace X with unikraft everywhere?

File: content/en/community/hackathons/sessions/baby-steps/content/Demos_01.-Building-and-Running-the-Helloworld-Application.md [hunk 4]
```diff
@@ -303,7 +313,7 @@ Our final binaries are located inside the `build/` directory.
 
 ##### Run
 
-Let's run the `linuxu` image by doing a Linux-like executable running:
+Let's run the `linuxu` image by doing a Linux-like executable running. `linuxu` stands for Linux userspace; a `linuxu` image is just an executable that will be launched as a process in the host system:
```
Comment (RaduNichita) [inline content/en/community/hackathons/sessions/baby-steps/content/Demos_01.-Building-and-Running-the-Helloworld-Application.md:316]:
```suggestion
Let's run the `linuxu` image by doing a Linux-like executable running. `linuxu` stands for Linux userspace; a `linuxu` image is just an executable that will be launched as a process on the host system:
```

File: content/en/community/hackathons/sessions/baby-steps/content/intro.md [hunk 5]
```diff
@@ -2,35 +2,35 @@ In this session we are going to understand the basic layout of the Unikraft work
...
-Unikraft is a unikernel SDK, meaning it offers you the blocks (source code, configuration and build system, runtime support) to build and run unikernels.
+Unikraft is a unikernel [SDK](https://en.wikipedia.org/wiki/Software_development_kit), meaning it offers you the blocks (source code, configuration and build system, runtime support) to build and run unikernels.
 A unikernel is a single image file that can be loaded and run as a separate running instance, most often a virtual machine.
 
 Summarily, Unikraft components are shown in the image below:
 
 ![arch selection menu](/docs/sessions/01-baby-steps/images/unikraft_components.png)
 
-Unikraft is the core component, consisting of core / internal libraries, the build system, and platform and architecture code.
+`Unikraft` is the core component, consisting of core / internal libraries (each providing a part of the functionality commonly found in an operating system), the build system, and platform and architecture code.
```
Comment (RaduNichita) [inline content/en/community/hackathons/sessions/baby-steps/content/intro.md:12]:
```suggestion
`Unikraft` is the core component, consisting of multiple core / internal libraries (each providing a part of the functionality commonly found in an operating system), the build system, and platform and architecture code.
```

File: content/en/community/hackathons/sessions/baby-steps/work/01-echo-back/main.c [hunk 6]
```diff
@@ -16,7 +16,10 @@ int receive_and_send(int client_recvfd, int client_sendfd)
     int bytes_received = 0;
 
     memset(buf, 0, BUFLEN);
+
+	// TODO
```
Comment (RaduNichita) [inline content/en/community/hackathons/sessions/baby-steps/work/01-echo-back/main.c:20]: I think there is wrong indentation here. And you can put the message from the bellow line on this one

```suggestion
	// TODO:  recv() buffer into buf;
```

File: content/en/community/hackathons/sessions/baby-steps/work/01-echo-back/main.c [hunk 7]
```diff
@@ -16,7 +16,10 @@ int receive_and_send(int client_recvfd, int client_sendfd)
...
     bytes_received = // recv() buffer into buf;
```
Comment (RaduNichita) [inline content/en/community/hackathons/sessions/baby-steps/work/01-echo-back/main.c:21]:
```suggestion
    bytes_received = 
```

File: content/en/community/hackathons/sessions/baby-steps/work/01-echo-back/main.c [hunk 8]
```diff
@@ -36,7 +39,9 @@ int main(int argc, char* argv[])
     int listenfd = -1;
     struct sockaddr_in serv_addr;
 
-    listenfd = // open socket() fd;
+    // TODO
```
Comment (RaduNichita) [inline content/en/community/hackathons/sessions/baby-steps/work/01-echo-back/main.c:42]:
```suggestion
    // TODO: open socket fd;
```

File: content/en/community/hackathons/sessions/baby-steps/work/01-echo-back/main.c [hunk 9]
```diff
@@ -36,7 +39,9 @@ int main(int argc, char* argv[])
...
+	listenfd = // open socket() fd;
```
Comment (RaduNichita) [inline content/en/community/hackathons/sessions/baby-steps/work/01-echo-back/main.c:43]:
```suggestion
	listenfd = 
```

File: content/en/community/hackathons/sessions/baby-steps/work/01-echo-back/main.c [hunk 10]
```diff
@@ -52,7 +57,9 @@ int main(int argc, char* argv[])
     serv_addr.sin_port = htons(LISTEN_PORT);
     serv_addr.sin_addr.s_addr = INADDR_ANY;
 
-    int err = // bind() the socket;
+    // TODO
```
Comment (RaduNichita) [inline content/en/community/hackathons/sessions/baby-steps/work/01-echo-back/main.c:60]:
```suggestion
    // TODO: bind() the socket;
```

File: content/en/community/hackathons/sessions/baby-steps/work/01-echo-back/main.c [hunk 11]
```diff
@@ -52,7 +57,9 @@ int main(int argc, char* argv[])
...
+	int err = // bind() the socket;
```
Comment (RaduNichita) [inline content/en/community/hackathons/sessions/baby-steps/work/01-echo-back/main.c:61]:
```suggestion
	int err = 
```

File: content/en/community/hackathons/sessions/baby-steps/work/01-echo-back/main.c [hunk 12]
```diff
@@ -66,7 +73,9 @@ int main(int argc, char* argv[])
 
     listen(listenfd, 1);
 
-    connfd = // accept() new connection;
+    // TODO
```
Comment (RaduNichita) [inline content/en/community/hackathons/sessions/baby-steps/work/01-echo-back/main.c:76]:
```suggestion
    // TODO: accept() new connection;
```

File: content/en/community/hackathons/sessions/baby-steps/work/01-echo-back/main.c [hunk 13]
```diff
@@ -66,7 +73,9 @@ int main(int argc, char* argv[])
...
+	connfd = // accept() new connection;
```
Comment (RaduNichita) [inline content/en/community/hackathons/sessions/baby-steps/work/01-echo-back/main.c:77]:
```suggestion
	connfd = 
```

File: content/en/community/hackathons/sessions/baby-steps/work/02-rot13/launch.sh [hunk 14]
```diff
@@ -1,24 +1,24 @@
 #!/bin/bash
 
 if test $# -ne 1; then
-    echo "Usage: $0 <kvm_image>" 1>&2
+    echo "Usage: $0 <path_to_KVM_IMAGE>" 1>&2
```
Comment (RaduNichita) [inline content/en/community/hackathons/sessions/baby-steps/work/02-rot13/launch.sh:4]:
```suggestion
    echo "Usage: $0 <path_to_kvm_image>" 1>&2
```

File: content/en/community/hackathons/sessions/baby-steps/work/04-store-strings/main.c [hunk 15]
```diff
@@ -22,9 +22,11 @@ int receive_and_send(int filefd, int client_recvfd, int client_sendfd)
 		exit(1);
 	}
 
-        // write to file before echoing back
+    // TODO
```
Comment (RaduNichita) [inline content/en/community/hackathons/sessions/baby-steps/work/04-store-strings/main.c:25]:
```suggestion
    // TODO: write to file before echoing back
```

File: content/en/community/hackathons/sessions/baby-steps/work/04-store-strings/main.c [hunk 16]
```diff
@@ -22,9 +22,11 @@ int receive_and_send(int filefd, int client_recvfd, int client_sendfd)
...
+	// write to file before echoing back
```
Comment (RaduNichita) [inline content/en/community/hackathons/sessions/baby-steps/work/04-store-strings/main.c:26]:
```suggestion
```

File: content/en/community/hackathons/sessions/baby-steps/work/04-store-strings/main.c [hunk 17]
```diff
@@ -22,9 +22,11 @@ int receive_and_send(int filefd, int client_recvfd, int client_sendfd)
...
 
-        // rot13
+    // TODO
```
Comment (RaduNichita) [inline content/en/community/hackathons/sessions/baby-steps/work/04-store-strings/main.c:28]:
```suggestion
    // TODO: apply rot13
```

File: content/en/community/hackathons/sessions/baby-steps/work/04-store-strings/main.c [hunk 18]
```diff
@@ -22,9 +22,11 @@ int receive_and_send(int filefd, int client_recvfd, int client_sendfd)
...
+	// rot13
```
Comment (RaduNichita) [inline content/en/community/hackathons/sessions/baby-steps/work/04-store-strings/main.c:29]:
```suggestion
```

File: content/en/community/hackathons/sessions/baby-steps/work/04-store-strings/main.c [hunk 19]
```diff
@@ -40,7 +42,9 @@ int main(int argc, char* argv[])
 	int listenfd = -1;
 	struct sockaddr_in serv_addr;
 
+	// TODO
```
Comment (RaduNichita) [inline content/en/community/hackathons/sessions/baby-steps/work/04-store-strings/main.c:45]:
```suggestion
	// TODO: open() the file you want to store the strings in;
```

File: content/en/community/hackathons/sessions/baby-steps/work/04-store-strings/main.c [hunk 20]
```diff
@@ -91,7 +95,8 @@ int main(int argc, char* argv[])
 
 	close(listenfd);
 
-        // close() filefd
+    // TODO
```
Comment (RaduNichita) [inline content/en/community/hackathons/sessions/baby-steps/work/04-store-strings/main.c:98]:
```suggestion
    // TODO: close() filefd
```

File: content/en/community/hackathons/sessions/baby-steps/work/04-store-strings/main.c [hunk 21]
```diff
@@ -91,7 +95,8 @@ int main(int argc, char* argv[])
...
+	// close() filefd
```
Comment (RaduNichita) [inline content/en/community/hackathons/sessions/baby-steps/work/04-store-strings/main.c:99]:
```suggestion
```

File: content/en/community/hackathons/sessions/baby-steps/work/02-rot13/launch.sh [hunk 22]
```diff
@@ -1,24 +1,24 @@
 #!/bin/bash
```
Comment (RaduNichita) [inline content/en/community/hackathons/sessions/baby-steps/work/02-rot13/launch.sh:1]: I suggest we have only one `launch.sh` file and make a symlink to it for the rest of the exercises?

File: content/en/community/hackathons/sessions/baby-steps/content/intro.md [hunk 23]
```diff
@@ -2,35 +2,35 @@ In this session we are going to understand the basic layout of the Unikraft work
 We are also going to take a look at how we can build basic applications and how we can extend their functionality and support by adding ported external libraries.
 
 Before everything, let's take a bird's eye view of what Unikraft is and what we can do with it.
-Unikraft is a unikernel SDK, meaning it offers you the blocks (source code, configuration and build system, runtime support) to build and run unikernels.
+Unikraft is a unikernel [SDK](https://en.wikipedia.org/wiki/Software_development_kit), meaning it offers you the blocks (source code, configuration and build system, runtime support) to build and run unikernels.
 A unikernel is a single image file that can be loaded and run as a separate running instance, most often a virtual machine.
 
 Summarily, Unikraft components are shown in the image below:
```
Comment (RaduNichita) [inline content/en/community/hackathons/sessions/baby-steps/content/intro.md:8]:
```suggestion
Summarily, the Unikraft components are shown in the image below:
```

File: content/en/community/hackathons/sessions/baby-steps/sol/04-store-strings/main.c [hunk 24]
```diff
@@ -55,7 +55,7 @@ int main(int argc, char* argv[])
     int listenfd = -1;
     struct sockaddr_in serv_addr;
 
-    int filefd = open("my_strings", O_WRONLY | O_CREAT | O_TRUNC, 0644);
+    int filefd = open("my_strings.txt", O_WRONLY | O_CREAT | O_TRUNC, 0644);
     if (filefd < 0) {
         fprintf(stderr, "socket");
```
Comment (RaduNichita) [inline content/en/community/hackathons/sessions/baby-steps/sol/04-store-strings/main.c:60]:
```suggestion
        fprintf(stderr, "open");
```

File: content/en/community/hackathons/sessions/baby-steps/sol/04-store-strings/main.c [hunk 25]
```diff
@@ -55,7 +55,7 @@ int main(int argc, char* argv[])
     int listenfd = -1;
     struct sockaddr_in serv_addr;
 
-    int filefd = open("my_strings", O_WRONLY | O_CREAT | O_TRUNC, 0644);
+    int filefd = open("my_strings.txt", O_WRONLY | O_CREAT | O_TRUNC, 0644);
```
Comment (RaduNichita) [inline content/en/community/hackathons/sessions/baby-steps/sol/04-store-strings/main.c:58]: I think we should rename `filefd` to `file_fd` and `listenfd` to `listen_fd` to match the rest of the code?

**Review by RaduNichita — CHANGES_REQUESTED** [review]
Good job, @eduardvintila !

I left some minor comments for the PR with some suggestions.I also think, if we were to follow the comment approach with `TODO: short explanation on the task`, the files for the `sol` folder should be modified too. So, when we run the command `diff -y work/main.c sol.main.c`, only the solution should be different

File: content/en/community/hackathons/sessions/baby-steps/work/02-rot13/launch.sh (hunk 22 above)
Comment (eduardvintila) [reply content/en/community/hackathons/sessions/baby-steps/work/02-rot13/launch.sh:1]: I think we should keep separate files for this. Students might alter the directory hierarchy when copying the materials.

File: content/en/community/hackathons/sessions/baby-steps/content/Demos_01.-Building-and-Running-the-Helloworld-Application.md (hunk 1 above)
Comment (eduardvintila) [reply content/en/community/hackathons/sessions/baby-steps/content/Demos_01.-Building-and-Running-the-Helloworld-Application.md:10]: The current naming convention seems to apply to all other sessions as well. I think for the moment we should keep it like that for consistency.

File: content/en/community/hackathons/sessions/baby-steps/sol/02-rot13/kraft.yaml [hunk 26]
```diff
@@ -0,0 +1,20 @@
+---
+specification: '0.5'
+name: 01-echo-back
+unikraft:
+  version: 0.10.0
```
Comment (RaduNichita) [inline content/en/community/hackathons/sessions/baby-steps/sol/02-rot13/kraft.yaml:5]:
```suggestion
  version: stable
```

File: content/en/community/hackathons/sessions/baby-steps/sol/02-rot13/kraft.yaml [hunk 27]
```diff
@@ -0,0 +1,20 @@
...
+targets:
+  - architecture: x86_64
+    platform: kvm
+libraries:
+  lwip:
+    version: 0.10.0
```
Comment (RaduNichita) [inline content/en/community/hackathons/sessions/baby-steps/sol/02-rot13/kraft.yaml:11]: I think if we need a lwip version, that should be added to the rest of the exercises too

Comment (eduardvintila) [reply content/en/community/hackathons/sessions/baby-steps/sol/02-rot13/kraft.yaml]: lwip can be safely set to `stable` as well

**Review by RaduNichita — APPROVED** [review]
LGTM now!

Thanks, @eduardvintila!
Reviewed-by: Radu Nichita [radunichita99@gmail.com](mailto:radunichita99@gmail.com])
Approved-by: Radu Nichita [radunichita99@gmail.com](mailto:radunichita99@gmail.com])

**Comment by StefanJum** [comment]
Hi @eduardvintila, nice work.
You will have to squash your commits.
Unless there are big specific changes, which is not the case here, we want only one commit, in order to keep the history as clean as possible.

**Comment by eduardvintila** [comment]
> Hi @eduardvintila, nice work. You will have to squash your commits. Unless there are big specific changes, which is not the case here, we want only one commit, in order to keep the history as clean as possible.

All done.

**Comment by StefanJum** [comment]
@eduardvintila you need to do another squash :)
@razvand, can you then take a look and see if everything is ok so we can merge it and do a rebuild for the website for today's session?
If you don't have time let me know and I can take a look.

**Comment by eduardvintila** [comment]
> @eduardvintila you need to do another squash :) @razvand, can you then take a look and see if everything is ok so we can merge it and do a rebuild for the website for today's session? If you don't have time let me know and I can take a look.

Yes, yes. Doing it now :)

**Pull request merged**